src/ddb/ddb_vos.c

```diff
@@ -0,0 +1,584 @@
+		args->sa_current++;
+		return 0;
+	}
+
+	switch (type) {
+	case VOS_ITER_COUUID:
+		uuid_copy(args->sa_uuid, entry->ie_couuid);
+		break;
+	case VOS_ITER_OBJ:
+		args->sa_uoid = entry->ie_oid;
+		break;
+	case VOS_ITER_DKEY:
+		args->sa_key = entry->ie_key;
+		break;
+	case VOS_ITER_AKEY:
+		args->sa_key = entry->ie_key;
+		break;
+	case VOS_ITER_SINGLE:
+		break;
+	case VOS_ITER_RECX:
+		args->sa_recx = entry->ie_orig_recx;
+		break;
+	case VOS_ITER_DTX:
+		break;
+	case VOS_ITER_NONE:
+		break;
+	}
+
+	return 1;
+}
+
+static int
+get_by_idx(daos_handle_t hdl, uint32_t idx, struct search_args *args, daos_unit_oid_t *uoid,
+	   daos_key_t *dkey, daos_key_t *akey, vos_iter_type_t type)
+{
+	vos_iter_param_t param = {0};
+	struct vos_iter_anchors anchors = {0};
+	int found;
+
+	args->sa_idx = idx;
+
+	param.ip_hdl = hdl;
+	param.ip_epr.epr_hi = DAOS_EPOCH_MAX;
+	if (uoid)
+		param.ip_oid = *uoid;
+	if (dkey)
+		param.ip_dkey = *dkey;
+	if (akey)
+		param.ip_akey = *akey;
+	found = vos_iterate(&param, type, false, &anchors, get_by_idx_cb, NULL, args, NULL);
+
+	if (!found)
+		return -DER_INVAL;
+
+	return 0;
+}
+
+int
+dv_get_cont_uuid(daos_handle_t poh, uint32_t idx, uuid_t uuid)
+{
+	struct search_args args = {0};
+	int rc;
+
+	rc = get_by_idx(poh, idx, &args, NULL, NULL, NULL, VOS_ITER_COUUID);
+	if (SUCCESS(rc))
+		uuid_copy(uuid, args.sa_uuid);
+	return rc;
+}
+
+int
+dv_get_object_oid(daos_handle_t coh, uint32_t idx, daos_unit_oid_t *uoid)
+{
+	struct search_args args = {0};
+	int rc;
+
+	D_ASSERT(uoid != NULL);
+	if (daos_handle_is_inval(coh))
+		return -DER_INVAL;
+
+	rc = get_by_idx(coh, idx, &args, NULL, NULL, NULL, VOS_ITER_OBJ);
+	if (SUCCESS(rc))
+		*uoid = args.sa_uoid;
+
+	return rc;
+}
+
+int
+dv_get_dkey(daos_handle_t coh, daos_unit_oid_t uoid, uint32_t idx, daos_key_t *dkey)
+{
+	struct search_args args = {0};
+	int rc;
+
+	D_ASSERT(dkey != NULL);
+	if (daos_handle_is_inval(coh) || daos_unit_oid_is_null(uoid))
+		return -DER_INVAL;
+
+	rc = get_by_idx(coh, idx, &args, &uoid, NULL, NULL, VOS_ITER_DKEY);
+	if (SUCCESS(rc))
+		*dkey = args.sa_key;
+
+	return rc;
+}
+
+int
+dv_get_akey(daos_handle_t coh, daos_unit_oid_t uoid, daos_key_t *dkey, uint32_t idx,
+	    daos_key_t *akey)
+{
+	struct search_args args = {0};
+	int rc;
+
+	D_ASSERT(dkey != NULL);
+	D_ASSERT(akey != NULL);
+	if (daos_handle_is_inval(coh) || daos_unit_oid_is_null(uoid))
+		return -DER_INVAL;
+
+	rc = get_by_idx(coh, idx, &args, &uoid, dkey, NULL, VOS_ITER_AKEY);
+	if (SUCCESS(rc))
+		*akey = args.sa_key;
+
+	return rc;
+}
+
+int
+dv_get_recx(daos_handle_t coh, daos_unit_oid_t uoid, daos_key_t *dkey, daos_key_t *akey,
+	    uint32_t idx, daos_recx_t *recx)
+{
+	struct search_args args = {0};
+	int rc;
+
+	D_ASSERT(dkey != NULL);
+	D_ASSERT(akey != NULL);
+	D_ASSERT(recx != NULL);
+	if (daos_handle_is_inval(coh) || daos_unit_oid_is_null(uoid))
+		return -DER_INVAL;
+
+	rc = get_by_idx(coh, idx, &args, &uoid, dkey, akey, VOS_ITER_RECX);
```

Comment:
This is fine for now but it would be more efficient if you simply do something like this at object level

struct indexes {
   int obj_idx;
   int dkey_idx;
   int akey_idx;
   int recx_idx;
}

Just pass this with argument to vos_iterate and set recursive = true.

In pre-call, if we are not at the index, just set *acts |= VOS_ITER_CB_SKIP;  This will immediately skip to next entry.
Then it will recurs to child (and it will automatically select single value or recx)
In post-call, set acts |= VOS_ITER_CB_ABORT.   This will stop iterator at current level.